src/controllers/plugins/Plugin.hpp

```diff
@@ -0,0 +1,238 @@
+                QString("license is not a string (its type is %1)").arg(type));
+        }
+
+        auto verObj = obj.value("version");
+        if (verObj.isString())
+        {
+            auto v =
+                semver::from_string_noexcept(verObj.toString().toStdString());
+            if (v.has_value())
+            {
+                this->version = v.value();
+            }
+            else
+            {
+                this->errors.emplace_back(
+                    "unable to parse version (use semver)");
+                this->version = semver::version(0, 0, 0);
+            }
+        }
+        else
+        {
+            auto type = QString::fromStdString(
+                std::string(magic_enum::enum_name(verObj.type())));
+            this->errors.emplace_back(
+                QString("version is not a string (its type is %1)").arg(type));
+            this->version = semver::version(0, 0, 0);
+        }
+        auto tagsObj = obj.value("tags");
+        if (!tagsObj.isUndefined())
+        {
+            if (!tagsObj.isArray())
+            {
+                auto type = QString::fromStdString(
+                    std::string(magic_enum::enum_name(licenseObj.type())));
+                this->errors.emplace_back(
+                    QString("tags is not an array (its type is %1)").arg(type));
+                return;
+            }
+
+            auto tagsArr = tagsObj.toArray();
+            for (int i = 0; i < tagsArr.size(); i++)
+            {
+                const auto &t = tagsArr.at(i);
+                if (!t.isString())
+                {
+                    this->errors.push_back(
+                        QString(
+                            "tags element #%1 is not a string (its type is %2)")
+                            .arg(i)
+                            .arg(QString::fromStdString(
+                                std::string(magic_enum::enum_name(t.type())))));
+                    return;
+                }
+                this->tags.push_back(t.toString());
+            }
+        }
+    }
+};
+
+class Plugin
+{
+public:
+    QString id;
+    PluginMeta meta;
+
+    Plugin(QString id, lua_State *state, PluginMeta meta,
+           const QDir &loadDirectory)
+        : id(std::move(id))
+        , meta(std::move(meta))
+        , loadDirectory_(loadDirectory)
+        , state_(state)
+    {
+    }
+
+    ~Plugin();
+
+    /**
+     * @brief Perform all necessary tasks to bind a command name to this plugin
+     * @param name name of the command to create
+     * @param functionName name of the function that should be called when the command is executed
+     * @return true if addition succeeded, false otherwise (for example because the command name is already taken)
+     */
+    bool registerCommand(const QString &name, const QString &functionName);
+
+    /**
+     * @brief Get names of all commands belonging to this plugin
+     */
+    std::set<QString> listRegisteredCommands();
```

Comment:
Does this have to be a `std::set`? Or would a `std::unordered_set`, `std::vector`, or `QList` be enough?

Reply:
Interesting, this conversion seems to be stable. I'll change it to be `std::unordered_set<QString>`